text/3594-expose-stackrealign-attribute.md

```diff
@@ -0,0 +1,85 @@
+- Feature Name: (`realign-stack-attr`)
+- Start Date: (2024-03-26)
+- RFC PR: [rust-lang/rfcs#3594](https://github.com/rust-lang/rfcs/pull/3594)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Expose the `alignstack` function attribute to rustc.
+
+# Motivation
+[motivation]: #motivation
+This is usefull when you have no guarantees about the alignment of the stack in an extern "C" function.
+Also verbatim from the attribute reference:
+Legacy x86 code uses 4-byte stack alignment. Newer aligned SSE instructions (like ‘movaps’) that work with the stack require operands to be 16-byte aligned. This attribute realigns the stack in the function prologue to make sure the stack can be used with SSE instructions.
+
+Interrupt service routines (ISRs) often require special treatment regarding stack alignment. When an interrupt occurs, the processor saves the current execution context onto the stack before transferring control to the ISR. However, the stack might not be aligned to the required boundary, especially in embedded systems where memory constraints are tight.
+
+By exposing a stack realignment feature to Rust, developers working on embedded systems or performance-critical applications gain the ability to guarantee stack alignment within ISRs and other critical code paths directly from Rust code. This not only simplifies development but also enhances the reliability and performance of systems that rely on interrupt handling.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+The `[realign_stack]` attribute can be added to a function to force the compiler to add alignment to that function.
+Usefull in cases where your code is called from a thread or a binary compiled with another compiler, that uses different aligmnet and thus lead to a corruption.
+
+```
+#[realign_stack]
+#[no_mangle]
+pub extern "C" fn callback_function() -> i32 {
+    println!("Called from callback!!");
+
+    0
+}
+```
+
+Also could be used like this if you need to specifiy the alignment boundary in bytes.

```

Comment:
What's a use-case for having to specify this? This is not motivated/explained.

Reply:
do you perhaps think we should drop the whole "specify alignment boundry" and focus on just adding the "no_stack_alignment_assumption" thing instead?

Reply:
If there's no motivation for specifying a particular alignment then yes I think that would be better. (No idea what the best name for such an attribute would be.)

Reply:
I will change it up a bit then